File: src/application-comparison.ts
/** Environment-attributed comparison of several evaluated alternatives for one
 * application entrypoint. Each candidate revision is measured by an ordinary
 * `algal.application-evaluation.v1` produced through the existing serial
 * incumbent-versus-candidate evaluator; this record joins those verdicts, the
 * shared frozen measurement set, and the host-declared environment label into
 * one retained, replayable comparison. The `selected` manifest must carry an
 * accepted verdict — a losing or incomplete alternative is retained evidence,
 * never a selection. The record grants no authority: activation still requires
 * the reproduced accepted evaluation, and restoration still requires the
 * explicit host restoration policy. */
import {
  applicationId, applicationJson, applicationList, applicationObject, applicationRef,
  applicationTag, getApplicationRecord, parseApplicationRevision, putApplicationRecord,
} from "./application-contract";
import {
  parseApplicationEvaluationRequest, verifyApplicationEvaluation,
  type AdaptationRuntime,
} from "./application-adaptation";
import { digestCanonical, type Digest } from "./digest";
import { canonicalize } from "./values";
import type { Store } from "./store";

export type ComparisonVerdict = "accepted" | "rejected" | "incomplete";
export type ApplicationComparisonResult = {
  /** Candidate revision evaluated against the incumbent. */
  revision: Digest;
  /** The manifest that revision selects for the compared entrypoint. */
  manifest: Digest;
  /** The reproduced `algal.application-evaluation.v1` record. */
  evaluation: Digest;
  verdict: ComparisonVerdict;
};
export type ApplicationComparison = {
  contract: "algal.application-comparison.v1";
  application: string;
  parentState: Digest;
  entrypoint: string;
  environment: string;
  cases: Digest;
  scorer: Digest;
  policy: Digest;
  /** Results sorted by candidate revision digest; revisions and manifests unique. */
  results: ApplicationComparisonResult[];
  /** The selected alternative's manifest, or null when nothing was accepted. */
  selected: Digest | null;
};

export const COMPARISON_LIMITS = Object.freeze({ results: 8 });

function parseResult(input: unknown): ApplicationComparisonResult {
  const v = applicationObject(input, ["revision", "manifest", "evaluation", "verdict"]);
  if (v.verdict !== "accepted" && v.verdict !== "rejected" && v.verdict !== "incomplete") throw new Error("Invalid comparison verdict");
  return { revision: applicationRef(v.revision), manifest: applicationRef(v.manifest), evaluation: applicationRef(v.evaluation), verdict: v.verdict };
}

export function parseApplicationComparison(input: unknown): ApplicationComparison {
  const v = applicationObject(input, ["contract", "application", "parentState", "entrypoint", "environment", "cases", "scorer", "policy", "results", "selected"]);
  applicationTag(v.contract, "algal.application-comparison.v1");
  const results = applicationList(v.results, COMPARISON_LIMITS.results, parseResult);
  if (!results.length) throw new Error("Comparison requires at least one result");
  if (results.some((r, i) => i > 0 && r.revision <= results[i - 1]!.revision)) throw new Error("Comparison results must be sorted by revision");
  if (new Set(results.map(r => r.manifest)).size !== results.length) throw new Error("Comparison manifests must be unique");
  if (new Set(results.map(r => r.evaluation)).size !== results.length) throw new Error("Comparison evaluations must be unique");
  const selected = v.selected === null ? null : applicationRef(v.selected);
  if (selected !== null && !results.some(r => r.manifest === selected && r.verdict === "accepted")) throw new Error("Comparison selection requires an accepted result");
  return {
    contract: "algal.application-comparison.v1",
    application: applicationId(v.application), parentState: applicationRef(v.parentState),
    entrypoint: applicationId(v.entrypoint), environment: applicationId(v.environment),
    cases: applicationRef(v.cases), scorer: applicationRef(v.scorer), policy: applicationRef(v.policy),
    results, selected,
  };
}

export type ProduceComparisonInput = {
  application: string;
  parentState: Digest;
  entrypoint: string;
  environment: string;
  /** Evaluation references measured under the same request fields; order is normalized away. */
  evaluations: Digest[];
  /** Manifest digest of the chosen alternative; requires an accepted verdict. */
  selected: Digest | null;
};

/** Recomputes every cited evaluation against the parent state and derives the
 * canonical comparison. Shared fields come from the evidence itself, so the
 * stored record is exactly what verification reproduces. */
async function deriveApplicationComparison(
  store: Store,
  input: { application: string; parentState: Digest; entrypoint: string; environment: string; evaluations: Digest[]; selected: Digest | null },
  runtime: AdaptationRuntime,
): Promise<ApplicationComparison> {
  const application = applicationId(input.application);
  const parentState = applicationRef(input.parentState);
  const entrypoint = applicationId(input.entrypoint);
  const environment = applicationId(input.environment);
  const evaluations = input.evaluations.map(applicationRef);
  if (!evaluations.length || evaluations.length > COMPARISON_LIMITS.results) throw new Error("Comparison evaluation bound exceeded");
  if (new Set(evaluations).size !== evaluations.length) throw new Error("Comparison evaluations must be unique");

  const results: ApplicationComparisonResult[] = [];
  let cases: Digest | null = null, scorer: Digest | null = null, policy: Digest | null = null;
  for (const evaluationRef of evaluations) {
    const checked = await verifyApplicationEvaluation(store, evaluationRef, parentState, runtime);
    const request = await getApplicationRecord(store, checked.evaluation.request, parseApplicationEvaluationRequest);
    if (request.parentState !== parentState || request.environment !== environment) {
      throw new Error("Comparison evaluation is not bound to this parent state and environment");
    }
    const candidate = await getApplicationRecord(store, request.candidateRevision, parseApplicationRevision);
    if (candidate.application !== application) throw new Error("Comparison candidate belongs to another application");
    const entry = candidate.entrypoints.find(e => e.name === entrypoint);
    if (!entry) throw new Error("Comparison entrypoint is not in the candidate revision");
    if (request.entrypoint !== entrypoint) throw new Error("Comparison evaluation measures another entrypoint");
    if (cases === null) { cases = request.cases; scorer = request.scorer; policy = request.policy; }
    else if (request.cases !== cases || request.scorer !== scorer || request.policy !== policy) {
      throw new Error("Comparison evaluations must share cases, scorer, and policy");
    }
    results.push({ revision: request.candidateRevision, manifest: entry.manifest, evaluation: evaluationRef, verdict: checked.verdict.status });
  }
  results.sort((a, b) => (a.revision < b.revision ? -1 : a.revision > b.revision ? 1 : 0));
  if (new Set(results.map(r => r.manifest)).size !== results.length) throw new Error("Comparison manifests must be unique");
  const selected = input.selected === null ? null : applicationRef(input.selected);
  if (selected !== null && !results.some(r => r.manifest === selected && r.verdict === "accepted")) {
    throw new Error("Comparison selection requires an accepted result");
  }
  return {
    contract: "algal.application-comparison.v1", application, parentState, entrypoint, environment,
    cases: cases!, scorer: scorer!, policy: policy!, results, selected,
  };
}

/** Derives and stores the comparison. All evidence is recomputed; a rejected
 * or incomplete verdict is retained but cannot be selected. */
export async function produceApplicationComparison(
  store: Store,
  input: ProduceComparisonInput,
  runtime: AdaptationRuntime,
): Promise<{ comparisonRef: Digest; comparison: ApplicationComparison }> {
  const comparison = await deriveApplicationComparison(store, input, runtime);
  return { comparisonRef: await putApplicationRecord(store, comparison), comparison };
}

/** Replays a stored comparison: every cited evaluation is re-verified against
 * the named parent state and the recomputed record must equal the stored one
 * byte-for-byte. */
export async function verifyApplicationComparison(
  store: Store,
  comparisonRef: Digest,
  expectedParentState: Digest,
  runtime: AdaptationRuntime,
): Promise<ApplicationComparison> {
  const stored = await getApplicationRecord(store, comparisonRef, parseApplicationComparison);
  if (stored.parentState !== applicationRef(expectedParentState)) throw new Error("Comparison parent state is stale");
  const recomputed = await deriveApplicationComparison(store, {
    application: stored.application, parentState: stored.parentState, entrypoint: stored.entrypoint,
    environment: stored.environment, evaluations: stored.results.map(r => r.evaluation), selected: stored.selected,
  }, runtime);
  if (canonicalize(applicationJson(recomputed)) !== canonicalize(applicationJson(stored))) {
    throw new Error("Comparison is not reproducible from its evidence");
  }
  return stored;
}

/** Binding check used by the default host when a transition cites a comparison
 * as evidence: the record must name this application and the commit's parent
 * state. Full evidence replay stays in `verifyApplicationComparison`. */
export function checkComparisonBinding(stored: ApplicationComparison, application: string, parentState: Digest): void {
  if (stored.application !== application || stored.parentState !== parentState) {
    throw new Error("Comparison evidence does not bind this transition");
  }
}

/** Stable identity for a comparison record (content digest). */
export function comparisonDigest(comparison: ApplicationComparison): Digest {
  return digestCanonical(applicationJson(comparison));
}

File: crates/algal/src/application_comparison.rs
//! Environment-attributed comparison of evaluated alternatives — port of
//! src/application-comparison.ts. Each row is reproduced through the ordinary
//! incumbent-versus-candidate evaluator; the record joins the verdicts, the
//! shared frozen measurement set, and the host-declared environment label.
//! `selected` requires an accepted verdict and grants no authority.

use serde_json::{Value, json};

use crate::application::parse_revision;
use crate::application_adaptation::{parse_evaluation_request, verify_application_evaluation};
use crate::application_memory::{
    app_id, app_json, app_object, app_ref, app_tag, get_record, opt_ref, put_record,
};
use crate::canonical::canonical;
use crate::contract::list;
use crate::effects::Host;
use crate::store::Store;
use crate::{Error, Result};

const COMPARISON_RESULT_LIMIT: usize = 8;

fn fail(message: &str) -> Error {
    Error::invalid(message)
}

fn parse_result(input: &Value) -> Result<()> {
    let v = app_object(input, &["revision", "manifest", "evaluation", "verdict"])?;
    app_ref(&v["revision"])?;
    app_ref(&v["manifest"])?;
    app_ref(&v["evaluation"])?;
    match v["verdict"].as_str() {
        Some("accepted") | Some("rejected") | Some("incomplete") => Ok(()),
        _ => Err(fail("Invalid comparison verdict")),
    }
}

/// `parseApplicationComparison` — closed record, sorted unique results, and a
/// selection that must name an accepted manifest.
pub fn parse_comparison(input: &Value) -> Result<Value> {
    let v = app_object(
        input,
        &[
            "contract",
            "application",
            "parentState",
            "entrypoint",
            "environment",
            "cases",
            "scorer",
            "policy",
            "results",
            "selected",
        ],
    )?;
    app_tag(&v["contract"], "algal.application-comparison.v1")?;
    app_id(&v["application"])?;
    app_ref(&v["parentState"])?;
    app_id(&v["entrypoint"])?;
    app_id(&v["environment"])?;
    app_ref(&v["cases"])?;
    app_ref(&v["scorer"])?;
    app_ref(&v["policy"])?;
    let results = v["results"]
        .as_array()
        .ok_or_else(|| fail("Comparison results must be a list"))?;
    if results.is_empty() || results.len() > COMPARISON_RESULT_LIMIT {
        return Err(fail("Comparison result bound exceeded"));
    }
    let mut revisions = Vec::with_capacity(results.len());
    let mut manifests = Vec::with_capacity(results.len());
    let mut evaluations = Vec::with_capacity(results.len());
    for row in results {
        parse_result(row)?;
        revisions.push(row["revision"].as_str().unwrap_or_default());
        manifests.push(row["manifest"].as_str().unwrap_or_default());
        evaluations.push(row["evaluation"].as_str().unwrap_or_default());
    }
    if revisions.windows(2).any(|pair| pair[0] >= pair[1]) {
        return Err(fail("Comparison results must be sorted by revision"));
    }
    let mut unique = manifests.clone();
    unique.sort_unstable();
    unique.dedup();
    if unique.len() != manifests.len() {
        return Err(fail("Comparison manifests must be unique"));
    }
    let mut unique = evaluations.clone();
    unique.sort_unstable();
    unique.dedup();
    if unique.len() != evaluations.len() {
        return Err(fail("Comparison evaluations must be unique"));
    }
    match &v["selected"] {
        Value::Null => {}
        selected => {
            let selected = app_ref(selected)?;
            let accepted = results.iter().any(|row| {
                row["manifest"].as_str() == Some(selected)
                    && row["verdict"].as_str() == Some("accepted")
            });
            if !accepted {
                return Err(fail("Comparison selection requires an accepted result"));
            }
        }
    }
    Ok(input.clone())
}

struct CompareRequest {
    application: String,
    parent_state: String,
    entrypoint: String,
    environment: String,
    evaluations: Vec<String>,
    selected: Option<String>,
}

fn parse_compare_request(input: &Value) -> Result<CompareRequest> {
    let v = app_object(
        input,
        &[
            "application",
            "parentState",
            "entrypoint",
            "environment",
            "evaluations",
            "selected",
        ],
    )?;
    let mut evaluations = Vec::new();
    for row in list(&v["evaluations"], COMPARISON_RESULT_LIMIT)? {
        evaluations.push(app_ref(row)?.to_owned());
    }
    Ok(CompareRequest {
        application: app_id(&v["application"])?.to_owned(),
        parent_state: app_ref(&v["parentState"])?.to_owned(),
        entrypoint: app_id(&v["entrypoint"])?.to_owned(),
        environment: app_id(&v["environment"])?.to_owned(),
        evaluations,
        selected: opt_ref(&v["selected"])?,
    })
}

/// Recomputes every cited evaluation against the parent state and derives the
/// canonical comparison — shared fields come from the evidence itself.
async fn derive_comparison(store: &Store, request: &CompareRequest, host: &Host) -> Result<Value> {
    if request.evaluations.is_empty() {
        return Err(fail("Comparison evaluation bound exceeded"));
    }
    let mut unique = request.evaluations.clone();
    unique.sort_unstable();
    unique.dedup();
    if unique.len() != request.evaluations.len() {
        return Err(fail("Comparison evaluations must be unique"));
    }
    let mut results = Vec::with_capacity(request.evaluations.len());
    let mut shared: Option<(String, String, String)> = None;
    for evaluation_ref in &request.evaluations {
        let evaluation =
            verify_application_evaluation(store, evaluation_ref, &request.parent_state, host)
                .await?;
        let evaluation_request = parse_evaluation_request(&get_record(
            store,
            evaluation["request"].as_str().unwrap_or(""),
        )?)?;
        if evaluation_request.parent_state != request.parent_state
            || evaluation_request.environment.as_deref() != Some(request.environment.as_str())
        {
            return Err(fail(
                "Comparison evaluation is not bound to this parent state and environment",
            ));
        }
        let candidate =
            parse_revision(&get_record(store, &evaluation_request.candidate_revision)?)?;
        if candidate.application != request.application {
            return Err(fail("Comparison candidate belongs to another application"));
        }
        if evaluation_request.entrypoint != request.entrypoint {
            return Err(fail("Comparison evaluation measures another entrypoint"));
        }
        let entry = candidate
            .entrypoints
            .iter()
            .find(|entry| entry.name == request.entrypoint)
            .ok_or_else(|| fail("Comparison entrypoint is not in the candidate revision"))?;
        let fields = (
            evaluation_request.cases.clone(),
            evaluation_request.scorer.clone(),
            evaluation_request.policy.clone(),
        );
        match &shared {
            None => shared = Some(fields),
            Some((c, s, p)) if c == &fields.0 && s == &fields.1 && p == &fields.2 => {}
            _ => {
                return Err(fail(
                    "Comparison evaluations must share cases, scorer, and policy",
                ));
            }
        }
        results.push(json!({
            "revision": evaluation_request.candidate_revision,
            "manifest": entry.manifest.clone(),
            "evaluation": evaluation_ref,
            "verdict": evaluation["verdict"]["status"].as_str().unwrap_or(""),
        }));
    }
    results.sort_by(|a, b| {
        a["revision"]
            .as_str()
            .unwrap_or_default()
            .cmp(b["revision"].as_str().unwrap_or_default())
    });
    let mut manifests: Vec<&str> = results
        .iter()
        .map(|row| row["manifest"].as_str().unwrap_or_default())
        .collect();
    manifests.sort_unstable();
    manifests.dedup();
    if manifests.len() != results.len() {
        return Err(fail("Comparison manifests must be unique"));
    }
    match &request.selected {
        None => {}
        Some(selected) => {
            let accepted = results.iter().any(|row| {
                row["manifest"].as_str() == Some(selected.as_str())
                    && row["verdict"].as_str() == Some("accepted")
            });
            if !accepted {
                return Err(fail("Comparison selection requires an accepted result"));
            }
        }
    }
    let (cases, scorer, policy) =
        shared.ok_or_else(|| fail("Comparison evaluation bound exceeded"))?;
    Ok(json!({
        "contract": "algal.application-comparison.v1",
        "application": request.application.as_str(),
        "parentState": request.parent_state.as_str(),
        "entrypoint": request.entrypoint.as_str(),
        "environment": request.environment.as_str(),
        "cases": cases,
        "scorer": scorer,
        "policy": policy,
        "results": results,
        "selected": request.selected.clone().map_or(Value::Null, Value::String),
    }))
}

/// `produceApplicationComparison` — derive and store the comparison.
pub async fn produce_comparison(
    store: &mut Store,
    input: &Value,
    host: &Host,
) -> Result<(String, Value)> {
    let request = parse_compare_request(input)?;
    let comparison = derive_comparison(store, &request, host).await?;
    let reference = put_record(store, &comparison)?;
    Ok((reference, comparison))
}

/// `verifyApplicationComparison` — re-verify every cited evaluation against
/// the named parent state and require byte-identical reproduction.
pub async fn verify_comparison(
    store: &Store,
    comparison_ref: &str,
    expected_parent_state: &str,
    host: &Host,
) -> Result<Value> {
    let stored = parse_comparison(&get_record(store, comparison_ref)?)?;
    if stored["parentState"].as_str() != Some(expected_parent_state) {
        return Err(fail("Comparison parent state is stale"));
    }
    let evaluations = stored["results"]
        .as_array()
        .map(|rows| {
            rows.iter()
                .filter_map(|row| row["evaluation"].as_str().map(str::to_owned))
                .collect()
        })
        .unwrap_or_default();
    let request = CompareRequest {
        application: stored["application"]
            .as_str()
            .unwrap_or_default()
            .to_owned(),
        parent_state: stored["parentState"]
            .as_str()
            .unwrap_or_default()
            .to_owned(),
        entrypoint: stored["entrypoint"].as_str().unwrap_or_default().to_owned(),
        environment: stored["environment"]
            .as_str()
            .unwrap_or_default()
            .to_owned(),
        evaluations,
        selected: opt_ref(&stored["selected"])?,
    };
    let recomputed = derive_comparison(store, &request, host).await?;
    if canonical(&app_json(&recomputed)?)? != canonical(&app_json(&stored)?)? {
        return Err(fail("Comparison is not reproducible from its evidence"));
    }
    Ok(stored)
}

/// `checkComparisonBinding` — the record must name this application and the
/// commit's parent state; full replay stays in `verify_comparison`.
pub fn check_comparison_binding(
    stored: &Value,
    application: &str,
    parent_state: &str,
) -> Result<()> {
    if stored["application"].as_str() != Some(application)
        || stored["parentState"].as_str() != Some(parent_state)
    {
        return Err(fail("Comparison evidence does not bind this transition"));
    }
    Ok(())
}
